supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigAsyncJobExecutionActor.scala

```diff
@@ -252,54 +250,45 @@ class DispatchedConfigAsyncJobExecutionActor(override val standardParams: Standa
     timeout
   }
 
   override def pollStatus(handle: StandardAsyncPendingExecutionHandle): SharedFileSystemRunState = {
-    handle.previousState match {
-      case None =>
+    (handle.previousState, exitCodeTimeout.flatMap(t => handle.previousState.map(_.expired(t)))) match {
+      case (None, _) =>
         // Is not set yet the status will be set default to running
         SharedFileSystemRunState("Running")
-      case Some(s) if (s.status == "Running" || s.status == "WaitingForReturnCode") && jobPaths.returnCode.exists =>
+      case (Some(s), _) if (s.status == "Running" || s.status == "WaitingForReturnCode") && jobPaths.returnCode.exists =>
         // If exitcode file does exists status will be set to Done always
         SharedFileSystemRunState("Done")
-      case Some(s) if s.status == "Running" =>
+      case (Some(s), Some(false)) => s // No action required if state is not yet expired
+      case (Some(s), Some(true)) if s.status == "Running" =>
         // Exitcode file does not exist at this point, checking is jobs is still alive
-        if (exitCodeTimeout.isEmpty) s
-        else if (isAlive(handle.pendingJob).fold({ e =>
-          log.error(e, s"Running '${checkAliveArgs(handle.pendingJob).argv.mkString(" ")}' did fail")
-          true
-        }, x => x)) s
-        else SharedFileSystemRunState("WaitingForReturnCode")
-      case Some(s) if s.status == "WaitingForReturnCode" =>
+        if (isAlive(handle.pendingJob).fold({ e =>
+              log.error(e, s"Running '${checkAliveArgs(handle.pendingJob).argv.mkString(" ")}' did fail")
+              true
+            }, x => x)) SharedFileSystemRunState("Running")
+            else SharedFileSystemRunState("WaitingForReturnCode")
+      case (Some(s), Some(true)) if s.status == "WaitingForReturnCode" =>
         // Can only enter this state when the exit code does not exist and the job is not alive anymore
         // `isAlive` is not called anymore from this point
 
         // If exit-code-timeout is set in the config cromwell will create a fake exitcode file with exitcode 137
-        exitCodeTimeout match {
-          case Some(timeout) =>
-            val currentDate = Calendar.getInstance()
-            currentDate.add(Calendar.SECOND, -timeout)
-            if (s.date.after(currentDate)) s
-            else {
-              jobLogger.error(s"Return file not found after $timeout seconds, assuming external kill")
-              val writer = new PrintWriter(jobPaths.returnCode.toFile)
-              // 137 does mean a external kill -9, this is a assumption but easy workaround for now
-              writer.println(9)
-              writer.close()
-              SharedFileSystemRunState("Failed")
-            }
-          case _ => s
-        }
-      case Some(s) if s.status == "Done" => s // Nothing to be done here
+        jobLogger.error(s"Return file not found after ${exitCodeTimeout.getOrElse("-")} seconds, assuming external kill")
```

Comment:
Just thinking out loud -- at this point, the exitCodeTimeout has to exist, is that right? We can't enter this case unless the job has exceeded the timeout? So theoretically one should never see an error that looks like: `Return file not found after - seconds, assuming external kill"`

Reply:
This is true when `exit-code-timeout-seconds` is not yet. Once this is set the value will become Some(true) if the timeout is passed, and so enter this case.
Under the right conditions and a job will never be lost this case will be never touched. Sadly not all system are that reliable ;)